examples/tv-casting-app/APIs.md

```diff
@@ -0,0 +1,680 @@
+[issue commands](#issuing-commands) to it, [read](#read-operations) current
+playback state, and [subscribe](#subscriptions) to playback events.
+
+### Issuing Commands
+
+_{Complete Command invocation examples:
+[Android](android/App/app/src/main/java/com/matter/casting/CommandInvocationExample.java)
+| [iOS](darwin/TvCasting/TvCasting/MTRCommandInvocationExampleViewModel.swift)}_
+
+To issue a command, the Casting Client can `contentAppEndpoint` on the
+CastingPlayer it connected to and handle the command response. It would first
+ensure the endpoint supports the required cluster/command.
+
+On Android, this can be implemented as follows:
+
+```java
+...
+if (contentAppEndpoint.get().hasCluster(ContentLauncher.class)) {
+    // get the Cluster to use
+    ContentLauncher contentLauncher = contentAppEndpoint.get().getCluster(ContentLauncher.class);
+
+    // invoke the command
+    CompletableFuture<ContentLauncher.LauncherResponse> contentLauncherResponseFuture =
+                contentLauncher.launchURL(
+		            new URL("https://YOUR_URL/id"), "displaystring", null);
+    		contentLauncherResponseFuture.thenApply(
+			(response) -> {
+				// handle the response
+				return null;
+			});
+}
+```
+
+On iOS, the commands can be sent as follows:
+
+```objectivec
+...
+if(endpoint.hasCluster(MTREndpointClusterTypeContentLauncher))
+{
+    let cluster: MTRContentLauncherCluster = endpoint.cluster(for: MTREndpointClusterTypeContentLauncher) as! MTRContentLauncherCluster
+    if(cluster.canLaunchURL())
+    {
+        cluster.launch(URL(string: "https://YOUR_URL/id")!) { launchResponse, err in
+            self.Log.info("invokeCommand completed with \(err)")
+            self.status = "ContentLauncher.LaunchURL completed!"
+
+            // consume the launchResponse
+        }
+    }
+    else
+    {
+        self.status = "Cannot launchURL on cluster"
+    }
+}
+else
+{
+    self.status = "No ContentLauncher cluster support found on the endpoint"
+}
+...
+```
+
+### Read Operations
+
+_{Complete Attribute Read examples:
+[Android](android/App/app/src/main/java/com/matter/casting/AttributeReadExample.java)
+| [iOS](darwin/TvCasting/TvCasting/MTRAttributeReadExampleViewModel.swift)}_
+
+The CastingClient may read an Attribute from the `contentAppEndpoint` on the
+CastingPlayer. It would ensure that the desired cluster / attribute are
+available for reading on the endpoint before trying to read it.
+
+On Android, an attribute can be read with implementation like the one below.
+
+```java
+// check if the Endpoint supports the Cluster to use
+if (contentAppEndpoint.get().hasCluster(MediaPlayback.class)) {
+    // get the Cluster to use
+    MediaPlayback mediaPlayback = contentAppEndpoint.get().getCluster(MediaPlayback.class);
+
+    Optional<Attribute<MediaPlayback.CurrentState>> currentStateAttribute = mediaPlayback.getCurrentState();
+    if(currentStateAttribute.isPresent() && currentStateAttribute.get().isAvailable())
```

Comment:
Is the need for both isPresent and isAvailable to differentiate between the media playback cluster server not setting the attribute versus explicitly setting it to null? Does the client care in this case? The reason is I'm wondering if Attribute.isAvailable is sufficient to handle both cases.

Reply:
In the Android APIs we are returning an Optional which is why I am making an isPresent check. Either we should return the Attribute itself instead of an Optional\<Attribute> - if we just want to check for isAvailable, or we check both, or we get rid of the isAvailable method.

Reply:
I think we should return the Attribute always and have the isAvailable() check in it. This way, clients can subscribe to it even if it's not available.

Reply:
Done. Returning an Attribute instead of the Optional\<Attribute\<>> and the Attribute has an isAvailable method that we are calling here.